test/framework/kubetest/run.go

```diff
@@ -0,0 +1,211 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubetest
+
+import (
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"os/user"
+	"path"
+	"runtime"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/cluster-api/test/framework"
+)
+
+const (
+	standardImage   = "us.gcr.io/k8s-artifacts-prod/conformance"
+	ciArtifactImage = "gcr.io/kubernetes-ci-images/conformance"
+)
+
+const (
+	DefaultGinkgoNodes            = 1
+	DefaultGinkoSlowSpecThreshold = 120
+)
+
+type RunInput struct {
+	// ClusterProxy is a clusterctl test framework proxy for the workload cluster
+	// for which to run kubetest against
+	ClusterProxy framework.ClusterProxy
+	// NumberOfNodes is the number of cluster nodes that exist for kubetest
+	// to be aware of
+	NumberOfNodes int
+	// UseCIArtifacts will fetch the latest build from the main branch of kubernetes/kubernetes
+	UseCIArtifacts bool
+	// ArtifactsDirectory is where conformance suite output will go
+	ArtifactsDirectory string
+	// Path to the kubetest e2e config file
+	ConfigFilePath string
+	// GinkgoNodes is the number of Ginkgo nodes to use
+	GinkgoNodes int
+	// GinkgoSlowSpecThreshold is time in s before spec is marked as slow
+	GinkgoSlowSpecThreshold int
+	// KubernetesVersion is the version of Kubernetes to test
+	KubernetesVersion string
```

Comment:
The version here should match with the cluster's Kubernetes version, right? Do we want to download a conformance image that matches with the cluster k8s version or conformance image version can be different than the cluster's k8s version?

Reply:
Added a server version discovery function which is consumed in Run so it's now optional. There may be a reason to run a later conformance suite on an older version, so have left the field in.